README.md
> [!WARNING]
>
> You are viewing a **__work-in-progress__** branch that is **constantly
> being rebased** and will likely be **eventually deleted**.
> **__Fork at your own risk__**.  The following text is also undergoing
> editing, i.e., _not everything mentioned below is strictly true_
> yet, as this is not yet any kind of release.

# LambdaMOO Server

This is the source for the latest version of the LambdaMOO server.

Veteran users with existing MOO databases will, at this point, want to
consult `./ChangeLog.txt`

## Getting Started

For those of you starting fresh, what to do will depend on how you got
here.  There are a number of posibilities:

### Building the LambdaMOO server from a distribution

This is arguably the simplest option for those less familiar with `git`
and the autotools.

Here, we assume you have downloaded a distribution tarball --- a file of
the form `LambdaMOO-1.9.nn-dist.tar.gz` --- from somewhere, did the
usual `tar xzf` to unpack it, `cd`ed into the directory you created,
and are now reading this file.

The process from here on is:

	./configure
	make
	make test

The `./configure` script may possibly complain about missing
pre-requisites.  Various situations are listed below.

### VPATH builds

To build in a different directory from where the sources live (useful
to simultaneously create versions with different options set), you
should instead do

	cd BUILD
	PATH/TO/SRC/configure --srcdir=PATH/TO/SRC
	make
	make test

where `BUILD` is a fresh directory and `PATH/TO/SRC` is whatever path
gets you from there back to the sources.  (E.g., you could just make a
new subdirectory of the source directory:

	mkdir b
	cd b
	../configure --srcdir=..
	make
	make test

).

Note that if you had previously done any building in the source
directory itself, you will need to do a `make distclean` there
before doing any VPATH builds.

### Building the LambdaMOO server from the git repository

This process is more involved:

	git clone https://github.com/wrog/lambdamoo SRCDIR
	cd SRCDIR
	autoconf
	./configure
	make
	make test

## Notes about prequisites that you might need

* `autoconf` is from the GNU Autotools suite.  Try the latest
   version first.  At the time of this writing, our package is known
   to work with autoconf versions in the range 2.69-2.72.

   Also __never__ run `autoreconf`.  We are not using the full suite.
   In particular, we are using neither `automake` nor `autoheader` and
   they will both get very confused.

* `gperf`, preferably version 3.1 or later

* A `yacc`-compatible parser generator: Any of `yacc`, `bison`, or
  `byacc` will do.  (We do not use `lex`.)

* If you are using the Unicode version of the server you will need
  `pcre`, the Perl Compatible Regular Expression library, by which we
  mean the old version, PCRE1, not the newer and fancier PCRE2.

  Confusingly, in Debian and Ubuntu, the package to install for
  this is actually labeled `libpcre3-dev`, I have no idea why.
  MacOS homebrew, FreeBSD, and Fedora all have perfectly
  reasonable `pcre` packages that do the right thing.

* `make`, preferably GNU make.


# Playing with options

`./configure --help=short`

will give you a summary of the configuration options available.

The last chapter of the Programmers Manual mentions some of the more
significant ones, but the full gory details are spelled out in
`options.h`

> [!NOTE]
>
> __To MOO veterans__:  Unlike with the pre-1.9.0
> server versions, you most likely __do not__ want to edit this file;
> there are now `./configure` flags that cover pretty much everything
> you are likely to need.

E.g.,

	./configure --enable-net=tcp,8888,out

sets all of `NETWORK_PROTOCOL=NP_TCP`, `DEFAULT_PORT=8888`, and
`OUTBOUND_NETWORK=OBN_ON` in a single argument.)


# Historical

See [`README.1997`](./README.1997) for Pavel Curtis' original
`README`, which has important information for people living in 1997
and may also be an interesting window on the world of 25 years ago for
the rest of us, what with its references to ancient operating systems,
FTP servers, mailing lists, and physical addresses that no longer
exist.


ChangeLog.txt
Version 1.0.0, 5 February 1991
-- First public release.

Version 1.1.0, 8 February 1991
-- recycle(x) now calls x:recycle() before destroying the object.
-- the little-used and functionally-dubious update() function has been removed.
   **** NOTE: Technically, this change should make this a major release, since
   ****       it can break old code that depended on update() existing.  In
   ****       practice, however, only $prog:@update used it (in LambdaCore.db,
   ****       anyway).  Thus, to install this server, you should first @rmverb
   ****       $prog:@update, since its code would otherwise be syntactically
   ****       illegal.
-- added renumber() and reset_max_object(), to facilitate automatically
   generating new versions of LambdaCore.db from the current LambdaMOO db.
-- made it so that only wizards can set the .name property of players; code in
   the database now checks that players keep unique names.
-- added set_player_flag() and boot_player() to enable registration-style
   player creation and @toad-ing.

Version 1.1,1, 12 February 1991
-- Much rearrangement of #include's to come as close as possible to conformance
   with ANSI C and POSIX.
-- Many changes to improve portability, especially to sites with only 'nearly'
   ANSI-compliant compilers.
-- Added backtrace-printing for tasks aborted for running too long.

Version 1.1.2, 13 February 1991
-- Fixed nasty infinite recursion problem on server panic.
-- Fixed the implementation of renumber() to really do the job.  It's now much
   more expensive (linear in the size of the database rather than constant
   time), but at least it's correct.  It's only intended for use on a tiny
   DB anyway.

Version 1.1.3, 11 March 1991
-- Fixed benign unparsing bug that always parenthesized unary-minus
   expressions.
-- Changed move() to insert the newly-moved object as the last item in the
   contents list, rather than the first.
-- Fixed a less-than-benign bug in the unparsing of conditional expressions,
   where 'x ? y | (z ? w | v)' was being unparsed without the parentheses.
-- Fixed a bug whereby an extra checkpoint timer was created every time the
   '.dump' command was used.
-- Added printing the value of 'this' to error backtraces whenever it is
   different from the verb location.
-- Fixed bug in propagation of property ownership on object creation.
-- WARNING: I believe that there is a storage leak somewhere in the server.  If
            so, it is not very fast.  With any luck, I'll find it soon.  For
            safety's sake, check on the size of your server's address space
            and reboot the server every few days if it's getting large.

Version 1.2.0, 11 April 1991
-- Fixed bug in parser that made 'after' a reserved word; this was part of an
   old (and never completely implemented) alternative syntax for the 'fork'
   statement.
-- Added memory_usage() function, to help in tracking down any memory-leak
   bugs, among other things.
-- Removed ancient restrictions on players recycling themselves.
-- Changed recycle() so that children of recycled objects are not themselves
   recycled.  Instead, they are adopted by their former grandparent.
-- Fixed bugs in which errors are returned from move() in certain cases.
-- Removed useless error value E_BADMOVE.
-- Fixed bug in move() that allowed the enterfunc to be called even if the
   object being moved was no longer valid.
-- Changed several uses of E_INVIND to E_INVARG; E_INVIND is now only used when
   an attempt is made to fetch or store the value of a property or verb from an
   invalid object.
-- Changed several functions to improve consistency in the order in which
   certain kinds of errors are checked for.
-- Made the allocator keep track of statistics concerning memory usage by type
   of object being allocated.
-- Changed queued_tasks() to return the value of 'this' in each task as well.
   This new value is returned at the end of the sublist for each task, so old
   code should continue to work unless it depended upon the exact length of the
   sublists.
-- Changed set_task_perms() not to return E_PERM if the argument is the same as
   the current permissions of the verb (i.e., if the change in permissions
   would be a no-op).
-- Made tonum() and toobj() more robust in the face of extra spaces before or
   after tokens.

Version 1.3.0, 7 June 1991
-- Removed the now-useless `.help' and `.mem' commands, and all code concerning
   the also-removed CHECKMALLOC option.
-- Fixed allocation bug in inheritance of 'dobjstr' and its ilk, when the
   current value of the variable has the wrong type (e.g., 'dobjstr' is a list
   when another verb-call happens).
-- Fixed the timers implementation not to try to free storage inside an
   interrupt procedure.
-- Changed move() so that :accept is called even if it's a wizard doing the
   moving; of course, the wizards gets to do the move even if :accept returns
   false.
-- Added 'off/off of' as a new preposition.
-- Fixed the built-in parser to return #-1 ($nothing) for the empty string,
   rather than #-2 ($ambiguous_match).
-- Fixed a few bugs in how the allocator keeps track of types.
-- Fixed two memory leaks:
	 1) If you returned or aborted from inside an iteration over a list,
	    the list was never freed.
	 2) If you killed a task, its associated environment was never freed.
-- Added ticks_left() and seconds_left(), for determining how much longer the
   current task will be allowed to run.
-- Reduced the maximum number of seconds for each task from 60 to 15.
-- Removed restriction that the first argument to notify() and boot_player()
   must be a valid player object.
-- Removed the .shutdown and .dump commands.
-- Added shutdown() and dump_database() functions to replace their respective
   built-in commands.  The shutdown() function takes a single string argument
   that is printed to all players as their connections are closed.
-- Added output_delimiters() as a way to discover the current PREFIX and SUFFIX

log.c
/******************************************************************************
  Copyright (c) 1992, 1995, 1996 Xerox Corporation.  All rights reserved.
  Portions of this code were written by Stephen White, aka ghond.
  Use and copying of this software and preparation of derivative works based
  upon this software are permitted.  Any distribution of this software or
  derivative works must comply with all applicable United States export
  control laws.  This software is made available AS IS, and Xerox Corporation
  makes no warranty about the software, its performance or its conformity to
  any specification.  Any person obtaining a copy of this software is requested
  to send their name and post office or electronic mail address to:
    Pavel Curtis
    Xerox PARC
    3333 Coyote Hill Rd.
    Palo Alto, CA 94304
    Pavel@Xerox.Com
 *****************************************************************************/

#include <errno.h>
#include "my-stdarg.h"
#include "my-stdio.h"
#include "my-string.h"
#include "my-time.h"

#include "bf_register.h"
#include "config.h"
#include "functions.h"
#include "log.h"
#include "options.h"
#include "storage.h"
#include "streams.h"
#include "utils.h"

static FILE *log_file = 0;

void
set_log_file(FILE * f)
{
    log_file = f;
}

int log_pcount = 5000;
static time_t log_prev = 0;
int log_report_progress_cktime(void)
{
    time_t now = time(0);
    log_pcount = 5000;
    return ((now >= log_prev + 2) && (log_prev = now, 1));
}

static void
do_log(const char *prefix, const char *fmt, va_list args)
{
    FILE *f;

    log_prev = time(0);
    log_pcount = 5000;
    if (log_file) {
	char *nowstr = ctime(&log_prev);

	nowstr[19] = '\0';	/* kill the year and newline at the end */
	f = log_file;
	fprintf(f, "%s: %s", nowstr + 4, prefix);	/* skip the day of week */
    } else
	f = stderr;

    vfprintf(f, fmt, args);
    fflush(f);
}

void
oklog(const char *fmt,...)
{
    va_list args;

    va_start(args, fmt);
    do_log("", fmt, args);
    va_end(args);
}

void
errlog(const char *fmt,...)
{
    va_list args;

    va_start(args, fmt);
    do_log("*** ", fmt, args);
    va_end(args);
}

void
log_perror(const char *what)
{
    errlog("%s: %s\n", what, strerror(errno));
}


#ifdef LOG_COMMANDS
static Stream *command_history = 0;
#endif

void
reset_command_history(void)
{
#ifdef LOG_COMMANDS
    if (command_history == 0)
	command_history = new_stream(1024);
    else
	reset_stream(command_history);
#endif
}

void
log_command_history(void)
{
#ifdef LOG_COMMANDS
    errlog("COMMAND HISTORY:\n%s", stream_contents(command_history));
#endif
}

void
add_command_to_history(Objid player, const char *command)
{
#ifdef LOG_COMMANDS
    time_t now = time(0);
    char *nowstr = ctime(&now);

    nowstr[19] = '\0';		/* kill the year and newline at the end */
    stream_printf(command_history, "%s: #%d: %s\n",
		  nowstr + 4,	/* skip day of week */
		  player, command);
#endif				/* LOG_COMMANDS */
}

/**** built in functions ****/

static package
bf_server_log(Var arglist, Byte next, void *vdata, Objid progr)
{
    if (!is_wizard(progr)) {
	free_var(arglist);
	return make_error_pack(E_PERM);
    } else {
	int is_error = (arglist.v.list[0].v.num == 2
			&& is_true(arglist.v.list[2]));

	if (is_error)
	    errlog("> %s\n", arglist.v.list[1].v.str);
	else
	    oklog("> %s\n", arglist.v.list[1].v.str);

	free_var(arglist);
	return no_var_pack();
    }
}

void
register_log(void)
{
    register_function("server_log", 1, 2, bf_server_log, TYPE_STR, TYPE_ANY);
}

char rcsid_log[] = "$Id$";

/*
 * $Log$
 * Revision 1.4  2004/05/22 01:25:43  wrog
 * merging in WROGUE changes (W_SRCIP, W_STARTUP, W_OOB)
 *
 * Revision 1.3.10.1  2003/06/03 12:19:27  wrog
 * added log_report_progress()
 *
 * Revision 1.3  1998/12/14 13:17:59  nop
 * Merge UNSAFE_OPTS (ref fixups); fix Log tag placement to fit CVS whims
 *
 * Revision 1.2  1997/03/03 04:18:48  nop
 * GNU Indent normalization
 *
 * Revision 1.1.1.1  1997/03/03 03:45:00  nop
 * LambdaMOO 1.8.0p5
 *
 * Revision 2.2  1996/04/08  01:06:25  pavel
 * Added `set_log_file()' entry point.  Made logging print undated messages to
 * stderr if they arrive before a log file has been set.  Release 1.8.0p3.
 *
 * Revision 2.1  1996/02/08  07:00:50  pavel
 * Renamed err/logf() to errlog/oklog().  Cleaned up bf_server_log() code.
 * Updated copyright notice for 1996.  Release 1.8.0beta1.
 *
 * Revision 2.0  1995/11/30  04:26:56  pavel
 * New baseline version, corresponding to release 1.8.0alpha1.
 *
 * Revision 1.10  1992/10/23  23:03:47  pavel
 * Added copyright notice.
 *
 * Revision 1.9  1992/10/21  03:02:35  pavel
 * Converted to use new automatic configuration system.
 *
 * Revision 1.8  1992/09/22  22:48:15  pavel
 * Added missing #include of "config.h".
 *
 * Revision 1.7  1992/09/08  22:04:45  pjames
 * changed `register_bf_log()' to `register_log()'
 *
 * Revision 1.6  1992/08/20  17:27:13  pavel
 * Added the prefix `> ' to all log messages generated by the server_log()
 * built-in function, so that they can be distinguished from server-generated
 * messages.
 *
 * Revision 1.5  1992/08/14  00:23:37  pavel
 * Commented text after #endif.
 *
 * Revision 1.4  1992/08/11  17:27:58  pjames
 * Changed server_log to return 0 instead of 1.
 *
 * Revision 1.3  1992/08/10  17:14:00  pjames
 * Updated #includes.  Added bf_server_log to write messages to server
 * log output.   Added registration procedure.
 *
 * Revision 1.2  1992/07/21  00:04:15  pavel
 * Added rcsid_<filename-root> declaration to hold the RCS ident. string.
 *
 * Revision 1.1  1992/07/20  23:23:12  pavel
 * Initial RCS-controlled version.
 */
